x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsampleStep.java

```diff
@@ -128,7 +131,7 @@ public void performAction(
 
     void performDownsampleIndex(String indexName, String downsampleIndexName, ActionListener<Void> listener) {
```

Comment:
Thanks for working on this Salvatore

I'm wondering, above at line `104` https://github.com/elastic/elasticsearch/pull/97557/files#diff-54833e304261a22efe3c0a185c58d4742391bd45ef654f1a0b729bcf3374a0e4R107
we seem to continue to fail the downsample operation and try again - the downsample index name 
will remain the same IIUC. Will this not continuously fail as the target downsample index exists already? 

Should we not check with the persistent task framework (or cluster state) if the operation already in progress?

Reply:
I think we should allow the transport downsample action to be invoked multiple times? If the target index already exists then we can skip that. If the persistent task already exists then we just need to wait for completion of the shard level rollup persistent tasks. 

The downsample action does perform a few post downsample operations like updating index settings, this happens via the elected master node. If ILM retries execution, then I think we should assume that the thread invoking downsample action failed too. So it should be allowed to execute downsample action multiple times.

Reply:
>I think we should allow the transport downsample action to be invoked multiple times? 

Oh sure that sounds good too. Is the TransportAction now treating a downsampling request for a target index that exists already as a no-op?

We have in `main` a race condition where we delete the downsample target index but RollupShardTask is still running and recreates the target index. So in ILM we rewind, delete the target index, generate a new one and try again. 

With this PR the target index name remains stable - is the whole rewind, delete, retry still needed?

Reply:
Coming back to some code above (not able to comment in this PR) 

But this bit  https://github.com/elastic/elasticsearch/pull/97557/files#diff-54833e304261a22efe3c0a185c58d4742391bd45ef654f1a0b729bcf3374a0e4R107 

I don't think we should fail the `DownsampleStep` (as ILM will retry it) if we detect that the target index exists and the downsampling is not complete (as the rollup task can still be running) - note this can also happen if the downsample operation timeouts on the client side (ILM side), not only on master failover.

I think this branch of failure should disappear completely in ILM and just call the downsample API again ? Then the transport downsampling action can identify that the rollup task is running and execute gracefully (i.e. notify the client, ILM, when downsampling completes). 

Also, just a quick question to consider - do we need to identify a complete failure of the rollup task vs a cancelled rollup task vs a time out in ILM and react differently to each of these? 
Ideally ILM should just retry downsampling and the transport downsample action should know how to treat a new incoming downsample request in each of these situations

Reply:
> Also, just a quick question to consider - do we need to identify a complete failure of the rollup task vs a cancelled rollup task vs a time out in ILM and react differently to each of these?
Ideally ILM should just retry downsampling and the transport downsample action should know how to treat a new incoming downsample request in each of these situations

A cancel will result in error being returned from the downsample action. ILM can then retry and downsample will start from where it left off. (the target index will still exist, only the persistent tasks need to be created. The downsample action will make sure that this happens)